kybra/compiler/kybra_generate/src/canister_method/post_upgrade_method/rust.rs

```diff
@@ -1,19 +1,112 @@
-use cdk_framework::{act::node::Param, traits::ToIdent};
+use cdk_framework::traits::CollectResults;
 use proc_macro2::TokenStream;
 use quote::quote;
+use rustpython_parser::ast::{Located, StmtKind};
 
-pub fn generate(params: &Vec<Param>) -> TokenStream {
-    let params_initializations = params.iter().map(|param| {
-        let param_name = format!("_cdk_user_defined_{}", param.name).to_ident();
-        let init_param_name = format!("INIT_PARAM_{}", param.name).to_ident();
+use super::generate_call;
+use crate::{source_map::SourceMapped, Error};
 
-        quote! {
-            #init_param_name.with(|x| *x.borrow_mut() = Some(#param_name));
+pub fn generate(
+    init_function_def_option: Option<&SourceMapped<&Located<StmtKind>>>,
+    post_upgrade_function_def_option: Option<&SourceMapped<&Located<StmtKind>>>,
+    entry_module_name: &String,
+) -> Result<TokenStream, Vec<Error>> {
+    let (call_to_init_py_function, call_to_post_upgrade_py_function) = (
+        generate_call(&init_function_def_option),
+        generate_call(&post_upgrade_function_def_option),
+    )
+        .collect_results()?;
+
+    let interpreter_init = generate_interpreter_init();
+    let ic_object_init = generate_ic_object_init();
+    let code_init = generate_code_init(entry_module_name);
+    let save_global_interpreter = generate_save_global_interpreter();
+    let call_to_user_init_or_post_upgrade = generate_call_to_user_init_or_post_upgrade(
+        &call_to_init_py_function,
+        &call_to_post_upgrade_py_function,
+    );
+
+    Ok(quote! {
+        let randomness = RANDOMNESS_STABLE_REF_CELL.with(|randomness_stable_ref_cell| randomness_stable_ref_cell.borrow().get().clone());
+
+        unsafe {
```

Comment:
I'm suspicious that this unsafe block actually applies to all of these and wonder if it's more prudent to tighten this down to surround only the things that are actually unsafe. As is, future devs would have to look through each method to see what is unsafe. It's convenient because they likely all access ref cells, but I think maybe we'd be better off doing the unsafe just around those accesses where possible.